### DISCUSSION TITLE
Can I enable `Member <...> is unknown` without enabling all of `reportGeneralTypeIssues`?

### DISCUSSION THREAD

**Discussion by Liam-DeVoe** [microsoft/pyright#6484] [Q&A] [answered]
errors reported by `reportGeneralTypeIssues` include reporting the following mistake, which I make all the time (via a typo or otherwise):

```python
instance = 0
instance.nonexistent_method()
# Cannot access member "nonexistent_method" for type "int"
#     Member nonexistent_method is unknown
```

and I would love to enable pyright in order to catch this. However, `reportGeneralTypeIssues` *also* includes reporting the following:

```python
def f(self, x: int):
    x = "a"
    # Expression of type "str" cannot be assigned to declared type "int"
    #     "str" is incompatible with "int"
```

which is also something that I do all the time in my code, and don't want to be reported.

For the record, I recognize that the latter is completely incorrect w.r.t. strong typing, and that pyright is correct to report it! But I abuse type annotations to do things like automatic deserialization (a la [pydantic](https://github.com/pydantic/pydantic)), and I *don't* always want to imply that the type should carry over to the method body itself. 

My question is: Can I enable the former, but disable the latter? Or equivalently, disable just the latter category from `reportGeneralTypeIssues`?

It seems a little weird that all of these unrelated type errors have been thrown into `reportGeneralTypeIssues`, even though pyright clearly has a finer-grained delineation internally; `Member <x> is unknown` vs `<x> is incompatible with <y>`. Even an option like `disableRegex` would help tremendously here.

**Comment by erictraut** [comment]
Thanks for posting. We've had requests in the past to make `reportGeneralTypeIssues` more fine-grained. [This](https://github.com/microsoft/pylance-release/issues/497) is the tracking issue.

I've said in the past that I'd be willing to consider splitting off a subset of the diagnostics into a separate diagnostic rule — but only if we can identify a subset that can be clearly identified and I can articulate a reason why someone might want to enable or disable this subset of diagnostics independently of others. In previous discussions, these requirements were not met. I think your articulation of the problem comes close to meeting these requirements.

The challenge here is that many of these errors are interrelated in Python. For example, the reason a "member cannot be accessed" could be any of 1) the class and its ancestors don't define a variable of that name, 2) the class defines an attribute of that name but it's a descriptor, and the `__get__` call is not allowed due to a call type violation, 3) the class defines a `__getattr__` method, but it can't be called due to a call type violation, 4) the access is a "set" or "delete" operation and the target attribute is read-only or `Final`, etc., etc.

I think it's worth considering moving all diagnostics related to attribute (member) accesses to a separate diagnostic rule. It could be called something like `reportAttributeAccessIssue`. I'll give this some thought.

That said, I'd encourage you to reconsider your approach to static type checking. The most fundamental job of a type checker is to tell you when you attempt to assign a value to a symbol that violates its declared type. In your code sample above, you've declared the type of `x` to be `int`, but you're assigning a `str` value to it. If you ignore or disable this fundamental capability of pyright, you're giving up on most of the value of a static type checker.

**Reply by Liam-DeVoe** [reply]
Thanks for the thoughtful response. 

I'm sympathetic to the position that there is a base level of value provided by pylance, and that if a user wants a strict subset of this functionality, they are not using the correct tool. In an ideal world, I think this would be the correct position to take. Unfortunately, we don't live in such a world.

I used to not use pylance's diagnostics at all. A equivalent (for my purposes) to the `Member <x> is unknown` diagnostic was caught by pylint's [E1101](https://pylint.readthedocs.io/en/latest/user_guide/messages/error/no-member.html). I recently switched from pylint to [ruff](https://github.com/astral-sh/ruff) due to linting speed, but ruff does not yet implement E1101. I thought I would give pylance a try to help fill in the gaps here, and ran into the above issue - because, as you alluded, I don't actually want the full capabilities of a basic type checker.

I understand why you may consider this an unsupported use case. I would only comment that supporting a finer-grained diagnostic disabling provides real utility to end users, even if it may not align with the platonic ideal of a type checker.

**Reply by erictraut** [reply, marked as answer]
> supporting a finer-grained diagnostic disabling provides real utility to end users

Perhaps, but only if it can be done in a way that makes sense. I'll also note that adding configuration knobs increases complexity of any tool (both for the user and the maintainer), so we don't add new knobs without a compelling reason. Plus, splitting this rule will be disruptive for pyright users who have existing configuration files. For these reasons, I'm going to reject the feature request for now. I think the cons outweigh the pros.

It sounds like you're looking for basic linting features rather than type checking, so pyright may not be the right tool for you.

I'll note that pylance's language server features (like completion suggestions, semantic highlighting, refactoring, etc.) are all based on static type information. If you use type annotations in your code, pylance will assume that you want these annotations to be honored. So, even if you disable the diagnostics related to static type checking, you'll still see odd behaviors in some of these language server features if you include annotations but then violate the type rules implied by these annotations.
